File: samplerxjavaapp/src/main/java/com/yoox/samplerxjavaapp/common/SearchResultItemAdapter.java
package com.yoox.samplerxjavaapp.common;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.bumptech.glide.Glide;
import com.yoox.net.models.outbound.SearchResultItem;
import com.yoox.samplerxjavaapp.R;

import java.util.ArrayList;
import java.util.List;

public class SearchResultItemAdapter extends RecyclerView.Adapter<SearchResultItemAdapter.ViewHolder> {

    private final List<SearchResultItem> data;
    private final OnItemSelectedListener onItemSelectedListener;

    public SearchResultItemAdapter(OnItemSelectedListener onItemSelectedListener) {
        this.data = new ArrayList<>(0);
        this.onItemSelectedListener = onItemSelectedListener;
    }

    public void replaceItems(List<SearchResultItem> newData) {
        data.clear();
        data.addAll(newData);
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        LayoutInflater inflater = LayoutInflater.from(context);

        View contactView = inflater.inflate(R.layout.item_search, parent, false);

        return new ViewHolder(contactView);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(data.get(position), onItemSelectedListener);
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    @Override
    public long getItemId(int position) {
        return data.get(position).hashCode();
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        private final ImageView image;
        private final TextView category;
        private final TextView brand;
        private final TextView price;

        ViewHolder(View itemView) {
            super(itemView);

            image = itemView.findViewById(R.id.image);
            category = itemView.findViewById(R.id.category);
            brand = itemView.findViewById(R.id.brand);
            price = itemView.findViewById(R.id.price);
        }

        void bind(SearchResultItem item, OnItemSelectedListener onItemSelectedListener) {

            Glide.with(image).load(item.getPreviewImage()).into(image);

            category.setText(item.getCategory().getName().getName());
            brand.setText(item.getBrand().getName());
            price.setText(item.getDiscountedPrice().getRawPrice());

            itemView.setOnClickListener(v -> onItemSelectedListener.onItemSelected(item));
        }
    }

    public interface OnItemSelectedListener {
        void onItemSelected(SearchResultItem item);
    }

}

File: samplekotlinapp/src/main/java/com/yoox/samplekotlinapp/common/SearchResultItemAdapter.kt
package com.yoox.samplekotlinapp.common

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.yoox.net.models.outbound.SearchResultItem
import com.yoox.samplekotlinapp.R

class SearchResultItemAdapter(
    private val onItemSelectedListener: (SearchResultItem) -> Unit
) :
    RecyclerView.Adapter<SearchResultItemAdapter.ViewHolder>() {

    private val data: MutableList<SearchResultItem> = mutableListOf()

    fun replaceItems(newData: List<SearchResultItem>) {
        data.clear()
        data.addAll(newData)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val context = parent.context
        val inflater = LayoutInflater.from(context)

        val contactView = inflater.inflate(R.layout.item_search, parent, false)

        return ViewHolder(contactView)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) =
        holder.bind(data[position], onItemSelectedListener)

    override fun getItemCount(): Int = data.size

    override fun getItemId(position: Int): Long = data[position].hashCode().toLong()

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val image: ImageView = itemView.findViewById(R.id.image)
        private val category: TextView = itemView.findViewById(R.id.category)
        private val brand: TextView = itemView.findViewById(R.id.brand)
        private val price: TextView = itemView.findViewById(R.id.price)

        fun bind(item: SearchResultItem, onItemSelectedListener: (SearchResultItem) -> Unit) {

            Glide.with(image).load(item.previewImage).into(image)

            category.text = item.category.name.name
            brand.text = item.brand.name
            price.text = item.discountedPrice.rawPrice

            itemView.setOnClickListener { onItemSelectedListener(item) }
        }
    }
}
